File: javascript/countdown/machine.js
const FrameEvent = require("./framelang/framelang");
// emitted from framec_v0.10.0
// get include files at https://github.com/frame-lang/frame-ancillary-files

class Countdown {
    
    // creating private properties
    
    #state
    #compartment
    #nextCompartment
    
    
    constructor (i) {
        
        // Create and intialize start state compartment.
        
        this.#state = this.#sS0_;
        this.#compartment = new CountdownCompartment(this.#state);
        this.#nextCompartment = null;
        this.#compartment.StateArgs["i"] = i;
        this.#compartment.StateVars["dec"] = 1;
        
        // Initialize domain
        
        // Send system start event
        const frameEvent = FrameEvent(">", null);
        this.#mux(frameEvent);
    }
    
    //====================== Multiplexer ====================//
    
    #mux(e) {
        switch (this.#compartment.state) {
            case this.#sS0_:
                this.#sS0_(e);
                break;
            case this.#sS1_:
                this.#sS1_(e);
                break;
            case this.#sStop_:
                this.#sStop_(e);
                break;
        }
        
        if( this.#nextCompartment != null) {
            let nextCompartment = this.#nextCompartment
            this.#nextCompartment = null
            if (nextCompartment._forwardEvent != null && 
               nextCompartment._forwardEvent._message == ">") {
                this.#mux(FrameEvent( "<", this.#compartment.ExitArgs))
                this.#compartment = nextCompartment
                this.#mux(nextCompartment._forwardEvent)
            } else {
                this.#doTransition(nextCompartment)
                if (nextCompartment._forwardEvent != null) {
                    this.#mux(nextCompartment._forwardEvent)
                }
            }
            nextCompartment._forwardEvent = null
        }
    }
    
    //===================== Machine Block ===================//
    
    #sS0_(e) {
        switch (e._message) {
            case ">":
                {
                (this.#compartment.StateArgs["i"]) = (this.#compartment.StateArgs["i"]) - (this.#compartment.StateVars["dec"]);
                this.print_do((this.#compartment.StateArgs["i"]).toString());
                if ((this.#compartment.StateArgs["i"]) == 0) {
                    let compartment =  new CountdownCompartment(this.#sStop_);
                    
                    
                    this.#transition(compartment);
                    return;
                }
                let compartment =  new CountdownCompartment(this.#sS1_);
                
                compartment.EnterArgs["i"] = this.#compartment.StateArgs["i"];
                
                this.#transition(compartment);
                return;
                }
                
        }
    }
    
    #sS1_(e) {
        switch (e._message) {
            case ">":
                {
                let compartment =  new CountdownCompartment(this.#sS0_);
                
                compartment.StateArgs["i"] = e._parameters["i"];
                compartment.StateVars["dec"] = 1;
                
                this.#transition(compartment);
                return;
                }
                
        }
    }
    
    #sStop_(e) {
        switch (e._message) {
            case ">":
                {
                this.print_do("done");
                return;
                }
                
        }
    }
    
    //===================== Actions Block ===================//
    
    print_do (s) {
        console.log(s)
    }
    
    // Unimplemented Actions
    
    
    //=============== Machinery and Mechanisms ==============//
    
    #transition(compartment) {
        this.#nextCompartment = compartment;
    }
    
    #doTransition(nextCompartment) {
        this.#mux(FrameEvent("<", this.#compartment.ExitArgs));
        this.#compartment = nextCompartment;
        this.#mux(FrameEvent(">", this.#compartment.EnterArgs));
    }
    
    
    
};

//=============== Compartment ==============//

class CountdownCompartment {

    constructor(state) {
        this.state = state
    }
    
    StateArgs = {};
    StateVars = {};
    EnterArgs = {};
    ExitArgs = {};
    _forwardEvent = FrameEvent.call(this)
}




class CountdownController extends Countdown {

	constructor(i) {
	  super(i)
	}
};

module.exports = CountdownController

File: python/countdown/machine.py
# emitted from framec_v0.10.0
# get include files at https://github.com/frame-lang/frame-ancillary-files
from framelang.framelang import FrameEvent

class Countdown:
    
    def __init__(self, i):
        
        # Create and intialize start state compartment.
        self.__state = self.__countdown_state_S0
        self.__compartment: 'CountdownCompartment' = CountdownCompartment(self.__state)
        self.__next_compartment: 'CountdownCompartment' = None
        self.__compartment.state_args["i"] = i
        self.__compartment.state_vars["dec"] = 1
        
        # Initialize domain
        
        # Send system start event
        frame_event = FrameEvent(">", None)
        self.__mux(frame_event)
    
    # ====================== Multiplexer ==================== #
    
    def __mux(self, e):
        if self.__compartment.state == self.__countdown_state_S0:
            self.__countdown_state_S0(e)
        elif self.__compartment.state == self.__countdown_state_S1:
            self.__countdown_state_S1(e)
        elif self.__compartment.state == self.__countdown_state_Stop:
            self.__countdown_state_Stop(e)
        
        if self.__next_compartment != None:
            next_compartment = self.__next_compartment
            self.__next_compartment = None
            if(next_compartment.forward_event is not None and 
               next_compartment.forward_event._message == ">"):
                self.__mux(FrameEvent( "<", self.__compartment.exit_args))
                self.__compartment = next_compartment
                self.__mux(next_compartment.forward_event)
            else:
                self.__do_transition(next_compartment)
                if next_compartment.forward_event is not None:
                    self.__mux(next_compartment.forward_event)
            next_compartment.forward_event = None
    
    # ===================== Machine Block =================== #
    
    def __countdown_state_S0(self, e):
        if e._message == ">":
            (self.__compartment.state_args["i"]) = (self.__compartment.state_args["i"]) - (self.__compartment.state_vars["dec"])
            self.print_do(str((self.__compartment.state_args["i"])))
            if  (self.__compartment.state_args["i"]) == 0:
                compartment = CountdownCompartment(self.__countdown_state_Stop)
                self.__transition(compartment)
                return
            
            compartment = CountdownCompartment(self.__countdown_state_S1)
            compartment.enter_args["i"] = self.__compartment.state_args["i"]
            self.__transition(compartment)
            return
        
    def __countdown_state_S1(self, e):
        if e._message == ">":
            compartment = CountdownCompartment(self.__countdown_state_S0)
            compartment.state_args["i"] = e._parameters["i"]
            compartment.state_vars["dec"] = 1
            self.__transition(compartment)
            return
        
    def __countdown_state_Stop(self, e):
        if e._message == ">":
            self.print_do("done")
            return
        
    
    # ===================== Actions Block =================== #
    
    
    def print_do(self,s: str):
        print(s)
    
    # Unimplemented Actions
    
    
    # =============== Machinery and Mechanisms ============== #
    
    def __transition(self, compartment: 'CountdownCompartment'):
        self.__next_compartment = compartment
    
    def  __do_transition(self, next_compartment: 'CountdownCompartment'):
        self.__mux(FrameEvent("<", self.__compartment.exit_args))
        self.__compartment = next_compartment
        self.__mux(FrameEvent(">", self.__compartment.enter_args))
    

# ===================== Compartment =================== #

class CountdownCompartment:

    def __init__(self, state):
        self.state = state
        self.state_args = {}
        self.state_vars = {}
        self.enter_args = {}
        self.exit_args = {}
        self.forward_event = FrameEvent(None, None)
    


# ********************

#class CountdownController(Countdown):
	#def __init__(self,i):
	    #super().__init__(i)

# ********************


